src/rgw/rgw_rest_s3.cc

```diff
@@ -4262,6 +4274,100 @@ rgw::auth::s3::LocalEngine::authenticate(
   return result_t::grant(std::move(apl), completer_factory(k.key));
 }
 
+/* OPAEngine */
+rgw::auth::Engine::result_t
+rgw::auth::s3::OPAEngine::authenticate(
+  const boost::string_view& _access_key_id,
+  const boost::string_view& signature,
+  const string_to_sign_t& string_to_sign,
+  const signature_factory_t& signature_factory,
+  const completer_factory_t& completer_factory,
+  const req_state* const s) const
+{
+  /* get the user info */
+  RGWUserInfo user_info;
+  const std::string access_key_id = _access_key_id.to_string();
+  if (rgw_get_user_info_by_access_key(store, access_key_id, user_info) < 0) {
+      ldout(cct, 5) << "error reading user info, uid=" << access_key_id
+              << " can't authenticate" << dendl;
+      return result_t::deny(-ERR_INVALID_ACCESS_KEY);
+  }
+
+  // Ask OPA for a decision
+  char buff[100];
+  snprintf(buff, sizeof(buff), "{\"input\":{\"method\":\"%s\", \"user\":\"%s\"}}", s->info.env->get("REQUEST_METHOD"), user_info.display_name.c_str());
+
+  dout(10) << "Request to OPA= " << std::string(buff) << dendl;
+
+  // Get OPA URL
+  char* opaUrl;
+  opaUrl = getenv("OPA_URL");
+  if (opaUrl == NULL)
+    opaUrl = (char *)"localhost:8181/v1/data/foo/blah";
+
+  dout(10) << "OPA URL= " << std::string(opaUrl) << dendl;
+
+  CURL *curl;
+  CURLcode res;
+
+  curl_global_init(CURL_GLOBAL_DEFAULT);
+  curl = curl_easy_init();
+  if(curl) {
+    std::string readBuffer;
+    struct curl_slist *headers = NULL;
+    headers = curl_slist_append(headers, "Content-Type: application/json");
+    curl_easy_setopt(curl, CURLOPT_URL, opaUrl);
+    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, buff);
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
+
+    res = curl_easy_perform(curl);
+    if(res != CURLE_OK)
+      fprintf(stderr, "curl_easy_perform() failed: %s\n",
+              curl_easy_strerror(res));
+
+    // Check OPA response
+    std::size_t found = readBuffer.find("true");
+    if (found == std::string::npos) {
+      ldout(cct, 2) << "OPA rejecting request" << dendl;
+      dout(10) << "failed to authorize request" << dendl;
+      return result_t::deny(-EPERM);
+    }
+
+    ldout(cct, 2) << "OPA accepting request" << dendl;
+
+    /* always cleanup */
+    curl_easy_cleanup(curl);
```

Comment:
Leaks?  We return prior if we succeed.   Or is the preceding comment not true?

Reply:
@pwmorreale This is an earlier version of the code. We are no longer using the curl library.